src/core/reference/include/openvino/reference/paged_attention.hpp

```diff
@@ -0,0 +1,330 @@
+    int32_t block_size = kv_cache_shape[2];
+    int32_t head_size = kv_cache_shape[3];
+
+    int32_t batch_tokens = q_shape[0];
+    int32_t query_features = q_shape[1];  // equals num_heads * head_size.
+    int32_t num_heads = query_features / head_size;
+    int32_t batch_seq = !past_lens_shape.empty() ? past_lens_shape[0] : 1;
+
+    // Process each query token.
+    for (int32_t token_idx = 0; token_idx < batch_tokens; token_idx++) {
+        // Determine sequence index using subsequence_begins if available.
+        int32_t seq_idx = 0;
+        if (batch_seq > 1 && subsequence_begins) {
+            for (int32_t s = 0; s < batch_seq; s++) {
+                if (token_idx >= subsequence_begins[s] && token_idx < subsequence_begins[s + 1]) {
+                    seq_idx = s;
+                    break;
+                }
+            }
+        }
+        // Process each head.
+        for (int32_t h = 0; h < num_heads; h++) {
+            const T* q_vec = query + token_idx * query_features + h * head_size;
+
+            int32_t seq_new_tokens =
+                subsequence_begins ? (subsequence_begins[seq_idx + 1] - subsequence_begins[seq_idx]) : 0;
+            int32_t seq_past_tokens = past_lens ? past_lens[seq_idx] : 0;
+            int32_t total_keys = seq_past_tokens + seq_new_tokens;
+            std::vector<T> scores(total_keys, T(0));
+
+            // Compute raw attention scores.
+            for (int32_t k = 0; k < total_keys; k++) {
+                T score_val = T(0);
+                if (k < seq_past_tokens) {
+                    // Retrieve key from cache.
+                    int32_t block_id = -1, token_offset = 0;
+                    bool found = paged_attention_utils::find_cached_token(seq_idx,
+                                                                          k,
+                                                                          block_indices,
+                                                                          block_indices_begins,
+                                                                          num_blocks,
+                                                                          block_size,
+                                                                          block_id,
+                                                                          token_offset);
+                    if (!found)
+                        continue;
+
+                    // If token falls within the sliding window of the first block, set score to -∞.
+                    int32_t first_block_for_seq =
+                        (block_indices_begins ? block_indices[block_indices_begins[seq_idx]] : -1);
+                    if (first_block_for_seq >= 0 && block_id == first_block_for_seq && token_offset < sliding_window) {
+                        score_val = -std::numeric_limits<T>::infinity();
+                    } else {
+                        int32_t kv_head = h % num_kv_heads;
```

Comment:
Should be `int32_t kv_head = h / (num_heads / num_kv_heads);` to support [MQA/GQA](https://github.com/openvinotoolkit/openvino/pull/21217#pullrequestreview-1747252695)?

Reply:
Right - after splitting from kv_head to k_head, v_head, now it is int32_t k_head = h / (num_q_heads / num_k_heads);